osu.Game/Online/Chat/DrawableLinkCompiler.cs

```diff
@@ -15,30 +19,133 @@ namespace osu.Game.Online.Chat
     /// <summary>
     /// An invisible drawable that brings multiple <see cref="Drawable"/> pieces together to form a consumable clickable link.
     /// </summary>
-    public class DrawableLinkCompiler : OsuHoverContainer
+    public class DrawableLinkCompiler : OsuHoverContainer, IHasContextMenu
     {
         /// <summary>
         /// Each word part of a chat link (split for word-wrap support).
         /// </summary>
         public List<Drawable> Parts;
 
+        [Resolved(CanBeNull = true)]
+        private OsuGame game { get; set; }
+
+        [Resolved]
+        private BeatmapManager beatmapManager { get; set; }
+
+        private readonly LinkDetails link;
+
         public override bool ReceivePositionalInputAt(Vector2 screenSpacePos) => Parts.Any(d => d.ReceivePositionalInputAt(screenSpacePos));
 
         protected override HoverClickSounds CreateHoverClickSounds(HoverSampleSet sampleSet) => new LinkHoverSounds(sampleSet, Parts);
 
-        public DrawableLinkCompiler(IEnumerable<Drawable> parts)
+        public DrawableLinkCompiler(IEnumerable<Drawable> parts, LinkDetails link)
         {
             Parts = parts.ToList();
+            this.link = link;
         }
 
-        [BackgroundDependencyLoader]
-        private void load(OsuColour colours)
+        [BackgroundDependencyLoader(true)]
+        private void load(OsuColour colours, OsuGame game, BeatmapManager beatmapManager)
         {
             IdleColour = colours.Blue;
+            this.game = game;
+            this.beatmapManager = beatmapManager;
         }
 
         protected override IEnumerable<Drawable> EffectTargets => Parts;
 
+        private BeatmapInfo getBeatmapFromLink(LinkDetails link)
+        {
+            var id = getBeatmapIdFromLink(link);
+            if (id.Type == StoreId.IdType.Beatmap)
+                return beatmapManager.QueryBeatmap(b => b.OnlineBeatmapID == id.Id);
+            else
+                return beatmapManager.QueryBeatmap(b => b.BeatmapSet.OnlineBeatmapSetID == id.Id);
+        }
+
+        private StoreId getBeatmapIdFromLink(LinkDetails link)
+        {
+            if (link.Action == LinkAction.OpenBeatmap && link.Argument != null && int.TryParse(link.Argument.Contains('?') ? link.Argument.Split('?')[0] : link.Argument, out int beatmapId))
+                return new StoreId(beatmapId, StoreId.IdType.Beatmap);
+            if (link.Action == LinkAction.OpenBeatmapSet && int.TryParse(link.Argument, out int setId))
+                return new StoreId(setId, StoreId.IdType.BeatmapSet);
+
+            return new StoreId(0, StoreId.IdType.Beatmap);
+        }
+
+        private class StoreId
```

Comment:
This class and the two methods above it are basically completely unnecessary. Why is there a "type" when you know the type of the link?

I'd put this all in the one switch below, or MAYBE lay this out as separate methods: `createBeatmapContextMenuItems()`, `createBeatmapSetContextMenuItems()`, `createProfileContextMenuItems()`, etc, though that may be over-engineered.